src/NHapi.Base/Parser/StructureDefinition.cs

```diff
@@ -110,24 +119,35 @@ public HashSet<string> GetNamesOfAllPossibleFollowingLeaves()
                 return namesOfAllPossibleFollowingLeaves;
             }
 
-            namesOfAllPossibleFollowingLeaves = new HashSet<string>();
-
-            var nextLeaf = NextLeaf;
-            if (nextLeaf != null)
+            // Double-Checked Locking
+            lock (allPossibleFollowingLeavesLock)
             {
-                namesOfAllPossibleFollowingLeaves.Add(nextLeaf.Name);
-                namesOfAllPossibleFollowingLeaves.UnionWith(nextLeaf.GetNamesOfAllPossibleFollowingLeaves());
-            }
+                if (namesOfAllPossibleFollowingLeaves != null)
+                {
+                    return namesOfAllPossibleFollowingLeaves;
+                }
 
-            var parent = Parent;
-            while (parent != null)
-            {
-                if (parent.IsRepeating)
+                var newNamesOfAllPossibleFollowingLeaves = new HashSet<string>();
+
+                var nextLeaf = NextLeaf;
+                if (nextLeaf != null)
+                {
+                    newNamesOfAllPossibleFollowingLeaves.Add(nextLeaf.Name);
+                    newNamesOfAllPossibleFollowingLeaves.UnionWith(nextLeaf.GetNamesOfAllPossibleFollowingLeaves());
+                }
+
+                var parent = Parent;
+                while (parent != null)
                 {
-                    namesOfAllPossibleFollowingLeaves.UnionWith(parent.GetAllPossibleFirstChildren());
+                    if (parent.IsRepeating)
+                    {
+                        newNamesOfAllPossibleFollowingLeaves.UnionWith(parent.GetAllPossibleFirstChildren());
+                    }
+
+                    parent = parent.Parent;
                 }
 
-                parent = parent.Parent;
+                namesOfAllPossibleFollowingLeaves = newNamesOfAllPossibleFollowingLeaves;
```

Comment:
Is this assignment atomic? If not, does that lead to a potential data race at the start of this method when `allPossibleFollowingLeavesLock` is not locked?

Reply:
It should be, access to the field is locked at the point of assignment by this method, the value is the last thing that is set before the lock is removed and the [volatile](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/volatile) keyword is used as per "[recommended practice](https://stackoverflow.com/questions/7210630/memory-model-guarantees-in-double-checked-locking/7230679#7230679)". That plus the lack of the JetBrains code inspection warning [Possible incorrect implementation of Double-Check Locking pattern. Read access to checked field](https://www.jetbrains.com/help/resharper/ReadAccessInDoubleCheckLocking.html) hopefully means that access to the field is synchronised correctly.